RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,332 @@
+# Improving subclassing Tensor by propagating subclass instances
+This RFC describes changes necessary to allow `__torch_function__` to be used
+by methods of `torch.Tensor` in an attempt to make subclassing more accessible
+to the users of the class. This entails making an API for subclass views
+public, and a change in the signature of `__torch_function__`.
+
+## Motivation and Scope
+Quoting [[1]], [[2]] and [[3]], the goals of this proposal are:
+
+1. Support subclassing `torch.Tensor` in Python
+2. Preserve `Tensor` subclasses when calling `torch` functions on them
+3. Use the PyTorch API with `torch.Tensor`-like objects that are _not_ `Tensor`
+   subclasses
+4. Preserve `Tensor` subclasses when calling `Tensor` methods.
+5. The ability to give external libraries a way to also define
+   functions/methods that follow the `__torch_function__` protocol.
+6. Propagating subclass instances correctly also with operators, using
+   views/slices/indexing/etc.
+7. Preserve subclass attributes when using methods or views/slices/indexing.
+8. A way to insert code that operates on both functions and methods uniformly
```

Comment:
This one is also separate from subclassing, so would also mark it separately as such. It is an oft-requested feature though. So there's 6 subclassing-related goals, then this goal, and then a maybe-goal.